File: C++/Disjoint Set.cpp
#include "stdafx.h"
//A disjoint set implementation permitting quick checks to see if two nodes are in the same set.
const int maxN = 10;
int id[maxN];

void init()
{
	for (int i = 0; i < maxN; i++)
		id[i] = i;
}

int root(int i)
{
	if (id[i] != i)
		return root(id[i]);

	return i;
}

void unite(int i, int j)
{
	id[root(i)] = root(j);
}

bool find(int i, int j)
{
	return root(i) == root(j);
}

#include <iostream>
int main()
{
	init();
	unite(1, 2);
	unite(4, 2);
	unite(3, 5);
	std::cout << find(1, 4) << "\n";
	std::cout << find(1, 3) << "\n";
	std::cout << find(3, 5) << "\n";
	unite(1, 5);
	std::cout << find(1, 3) << "\n";
    return 0;
}

File: Algorithms-Data-Structures/C#/Disjoint Set.cs
public class DisjointSet
{
    public readonly int[] DS;

    public DisjointSet(int capacity)
    {
        DS = new int[capacity];
        for (int i = 0; i < capacity; i++)
            DS[i] = i;
    }

    public int Root(int i)
    {
        if (DS[i] != i)
            DS[i] = Root(DS[i]);
        return DS[i];
    }

    public bool Find(int i, int j)
    {
        return Root(i) == Root(j);
    }

    public void Unite(int i, int j)
    {
        DS[Root(i)] = Root(j);
    }
}

